Geometry/VeryForwardGeometry/interface/CTPPSPixelSimTopology.h

```diff
@@ -4,209 +4,222 @@
-      x = x + simX_width_/2.;
- 
-      // now x in the system centered in the bottom left corner of the sensor (sensor view, rocs behind)
-      if ( x < 0. || x > simX_width_ )
-        throw cms::Exception("CTPPSPixelSimTopology")<< "out of reference frame";
-
-      // rows (x segmentation)
-      unsigned int arow;
-      if      ( x <= ( dead_edge_width_+    pitch_simX_ ) ) arow = 0;
-      else if ( x <= ( dead_edge_width_+ 79*pitch_simX_ ) ) arow = int( ( x-dead_edge_width_-pitch_simX_ )/pitch_simX_ )+1;
-      else if ( x <= ( dead_edge_width_+ 81*pitch_simX_ ) ) arow = 79;
-      else if ( x <= ( dead_edge_width_+ 83*pitch_simX_ ) ) arow = 80;
-      else if ( x <= ( dead_edge_width_+162*pitch_simX_ ) ) arow = int( ( x-dead_edge_width_-pitch_simX_ )/pitch_simX_ )-1;
-      else arow = 159;
-
-      arow = 159-arow;
-      if ( arow>159 )
-        throw cms::Exception("CTPPSPixelSimTopology")<< "row number exceeding limit";
-
-      return arow;
-    }
-
-    inline unsigned int col( double y ) const {
-      // y in the G4 simulation system
-      unsigned int column;
-
-      // columns (y segmentation)
-      // now y in the system centered in the bottom left corner of the sensor (sensor view, rocs behind)
-      y = y + simY_width_/2.;
-      if ( y < 0. || y > simY_width_ )
-        throw cms::Exception("CTPPSPixelSimTopology")<< " out of reference frame";
-
-      if      ( y <= ( dead_edge_width_+    pitch_simY_ ) ) column = 0;
-      else if ( y <= ( dead_edge_width_+ 51*pitch_simY_ ) ) column = int( ( y-dead_edge_width_-pitch_simY_ )/pitch_simY_ )+1;
-      else if ( y <= ( dead_edge_width_+ 53*pitch_simY_ ) ) column = 51;
-      else if ( y <= ( dead_edge_width_+ 55*pitch_simY_ ) ) column = 52;
-      else if ( y <= ( dead_edge_width_+105*pitch_simY_ ) ) column = int( ( y-dead_edge_width_-pitch_simY_ )/pitch_simY_ )-1;
-      else if ( y <= ( dead_edge_width_+107*pitch_simY_ ) ) column = 103;
-      else if ( y <= ( dead_edge_width_+109*pitch_simY_ ) ) column = 104;
-      else if ( y <= ( dead_edge_width_+159*pitch_simY_ ) ) column = int( ( y-dead_edge_width_-pitch_simY_ )/pitch_simY_ )-3;
-      else column = 155;
-
-      return column;
+    double lower_simX_border_;
+    double higher_simX_border_;
+    double lower_simY_border_;
+    double higher_simY_border_;
+    double eff_factor_;
+    unsigned short pixel_row_no_;
+    unsigned short pixel_col_no_;
+    unsigned short pixel_index_;
+  };
+
+public:
+  CTPPSPixelSimTopology();
+  ~CTPPSPixelSimTopology() {}
+
+  PixelInfo getPixelsInvolved(double x, double y, double sigma, double& hit_pos_x, double& hit_pos_y) const;
+
+  inline void pixelRange(unsigned int arow,
+                         unsigned int acol,
+                         double& lower_x,
+                         double& higher_x,
+                         double& lower_y,
+                         double& higher_y) const {
+    // x and y in the system  of Geant4 SIMULATION
+    arow = 159 - arow;
```

Comment:
This class seems to be full of magic numbers. These are not new in this PR, but at some point they should be replaced with named constants (or even better, obtained from geometry XML specifications in case any of them changes in the future).